scipy/special/boost_special_functions.h

```diff
@@ -760,46 +760,86 @@ ncf_pdf_double(double x, double v1, double v2, double l)
 
 template<typename Real>
 Real
-ncf_cdf_wrap(const Real x, const Real v1, const Real v2, const Real l)
+ncf_cdf_wrap(const Real v1, const Real v2, const Real l, const Real x)
 {
-    if (std::isfinite(x)) {
-        return boost::math::cdf(
-            boost::math::non_central_f_distribution<Real, StatsPolicy>(v1, v2, l), x);
+    if (std::isnan(x) || std::isnan(v1) || std::isnan(v2) || std::isnan(l)) {
+	return NAN;
     }
-    // -inf => 0, inf => 1
-    return 1.0 - std::signbit(x);
+    if ((v1 <= 0) || (v2 <= 0) || (l < 0)) {
+	sf_error("ncfdtr", SF_ERROR_DOMAIN, NULL);
+	return NAN;
+    }
+    if (std::isinf(x)) {
+	// -inf => 0, inf => 1
```

Comment:
To be self-consistent and consistent with previous behavior, this should return NaN for all `x < 0`, right?

Reply:
Ah, good point. That trips me up too. I'll change it. I think I understand the rationale at least. For example, ifwe defined `betainc` to be the CDF of the Beta distribution, then it's natural to say `betainc(2, 2, -0.2)` is zero, since if $X \sim B(2, 2)$, then $P(X < -0.2) = 0$. Treating it just as a special function though, the natural way to extend things is by analytic continuation, and if we haven't implemented the analytic continuation (or if we're looking at the real valued version and the analytic continuation on the real axis but outside of the usual domain for the distribution function returns complex values, then it makes sense to return `NAN`.

Reply:
> For example, if we defined betainc to be the CDF of the Beta distribution

I think that's my point, though: `betainc` *isn't* the CDF of the beta distribution; `btdtr` is / was. So I would have extended `betainc`'s domain or returned NaN, whereas I would think `btdtr` would only be for reals and would return 0 or 1 outside the distribution's domain.

Reply:
Yeah, I agree it's weird that we have different versions of the some functions, one a CDF, the other not, but the CDF doesn't behave as one would expect. I would imagine it's just an historical thing. It seems like the consistent principle being followed may be "whether something says it's a CDF or not, we don't treat it as a CDF, and the only valid way to extend a domain is by analytic continuation"?

Would you support maintaining CDF and non-CDF versions of the same functions with the only difference being how they treat arguments outside of the usual domain? It seems like a reasonable thing to do.

Reply:
> It seems like the consistent principle being followed may be "whether something says it's a CDF or not, we don't treat it as a CDF, and the only valid way to extend a domain is by analytic continuation"?

I think that describes what is happening. And with that in mind, yes, I imagine the reason for the apparent duplication is just historical.

> Would you support maintaining CDF and non-CDF versions of the same functions with the only difference being how they treat arguments outside of the usual domain?

Personally, I wouldn't want to work toward it because I think the random variable infrastructure can serve the need for functions to be treated as CDFs. On one hand, it would be great if the infrastructure could just call the special function and it would behave as a CDF, etc. There is lower overhead if this behavior happens at the compiled level. But if there is a need for this sort of behavior to be added by the distribution infrastructure (e.g. user-defined distributions) it might as well be applied consistently to all distributions, and then there is not much need to duplicate it in `special`.